### PR TITLE
Add a tutorial page about user-defined sampler.

### PR DESCRIPTION
**Pull request by sile**
This PR adds a tutorial page that describes Optuna's user-defined sampler feature.

### CODE REVIEW

File: docs/source/tutorial/sampler.rst [hunk 1]
```diff
@@ -0,0 +1,123 @@
+.. _sampler:
+
+User-Defined Sampler
+=====================
```
Comment (toshihikoyanase) [inline docs/source/tutorial/sampler.rst]: I think this line has too many `=`. Please remove the last one.

File: docs/source/tutorial/sampler.rst [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+
+Optuna allows users to create user-defined samplers.
+
+A sampler has the responsibility to determine the parameter values to be evaluated in a trial.
+When a `suggest` API (e.g., :func:`~optuna.trial.Trial.suggest_uniform`) is called inside an objective function, the corresponding distribution object (e.g., :class:`~optuna.distributions.UniformDistribution`) is created internally. A sampler samples a value from the distribution. The sampled value is returned to the caller of the `suggest` API and evaluated in the objective function.
+
+Optuna provides built-in samplers (e.g., :class:`~optuna.samplers.TPESampler`, :class:`~optuna.samplers.RandomSampler`) that work well for a wide range of cases.
+However, if you are only interested in optimizing hyperparameters in a specific domain, optimization performance may be improved if you use a sampling algorithm specialized to the domain.
```
Comment (toshihikoyanase) [inline docs/source/tutorial/sampler.rst]: This sentence contains two if-clauses. Could you simplify it to improve the readability?
For example, I think we can remove the first if-clause as follows:
> However, optimization performance may be improved if you use a sampling algorithm specialized for your problem.

File: docs/source/tutorial/sampler.rst [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+An Example: Implementing SimulatedAnnealingSampler
+--------------------------------------------------
+
+For creating a new sampler, you need to define a class that inherits :class:`~optuna.samplers.BaseSampler`,
+and implement the three abstract methods of the base class (
+:meth:`~optuna.samplers.BaseSampler.infer_relative_search_space`,
+:meth:`~optuna.samplers.BaseSampler.sample_relative` and
+:meth:`~optuna.samplers.BaseSampler.sample_independent`).
+
+As an example, the following code defines a sampler named ``SimulatedAnnealingSampler`` that based on
```
Comment (toshihikoyanase) [inline docs/source/tutorial/sampler.rst]: `that based on` seems to be a typo for `that is based on`.

**Comment by sile** [comment]
@toshihikoyanase Thank you for your review. I updated the documentation for reflecting your comments (https://github.com/pfnet/optuna/pull/451/commits/308707bc62f607991c7d152bb6529bd5e9404fc1).

**Pull request merged**